hack/generator/pkg/codegen/code_generator.go

```diff
@@ -104,8 +115,94 @@ func (generator *CodeGenerator) Generate(ctx context.Context, outputFolder strin
 	return nil
 }
 
-func (generator *CodeGenerator) CreatePackagesForDefinitions(definitions []astmodel.TypeDefiner) ([]*astmodel.PackageDefinition, error) {
-	packages := make(map[astmodel.PackageReference]*astmodel.PackageDefinition)
+func (generator *CodeGenerator) MarkLatestResourceVersionsForStorage(pkgs []*astmodel.PackageDefinition) ([]*astmodel.PackageDefinition, error) {
+
+	var result []*astmodel.PackageDefinition
+
+	resourceLookup, err := groupResourcesByVersion(pkgs)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, pkg := range pkgs {
+
+		resultPkg := astmodel.NewPackageDefinition(pkg.GroupName, pkg.PackageName)
+		for _, def := range pkg.Definitions() {
+			if structDef, ok := def.(*astmodel.StructDefinition); ok && structDef.IsResource() {
+
+				unversionedName, err := getUnversionedName(structDef.TypeName)
+				if err != nil {
+					// should never happen as all resources have versioned names
+					return nil, err
+				}
+
+				allVersionsOfResource := resourceLookup[unversionedName]
+				latestVersionOfResource := allVersionsOfResource[len(allVersionsOfResource)-1]
+
+				isLatestVersion := structDef.Name().PackageReference.PackagePath() == latestVersionOfResource.Name().PackageReference.PackagePath()
+				structDef = structDef.WithIsStorageVersion(isLatestVersion)
+
+				resultPkg.AddDefinition(structDef)
+			} else {
+				resultPkg.AddDefinition(def)
+			}
+		}
+
+		result = append(result, resultPkg)
+	}
+
+	return result, nil
+}
+
+func getUnversionedName(name *astmodel.TypeName) (unversionedName, error) {
+
+	group, _, err := name.PackageReference.GroupAndPackage()
+	if err != nil {
+		return unversionedName{}, err
+	}
+
+	return unversionedName{group, name.Name()}, nil
+}
+
+type unversionedName struct {
+	group string
+	name  string
+}
+
+func groupResourcesByVersion(pkgs []*astmodel.PackageDefinition) (map[unversionedName][]*astmodel.StructDefinition, error) {
+	result := make(map[unversionedName][]*astmodel.StructDefinition)
+
+	for _, pkg := range pkgs {
+		for _, def := range pkg.Definitions() {
+			if structDef, ok := def.(*astmodel.StructDefinition); ok && structDef.IsResource() {
+				name, err := getUnversionedName(structDef.TypeName)
+				if err != nil {
+					// this should never happen as resources will all have versioned names
+					return nil, fmt.Errorf("Unable to extract unversioned name in groupResources: %w", err)
+				}
+
+				if list, ok := result[name]; ok {
```

Comment:
I think that this block could be rewritten:
```
list := result[name]
result[name] = append(list, structDef)
```

because the 0 value of `[]*astmodel.StructDefinition` is `nil`

Reply:
`result[name] = append(result[name], structDef)` 🙂